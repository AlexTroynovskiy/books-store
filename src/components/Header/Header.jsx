import React from 'react';
import { NavLink } from 'react-router-dom';
import PropTypes from 'prop-types';

import { Container, Grid, AppBar, Toolbar, Typography, Badge } from '@material-ui/core';
import AddShoppingCartIcon from '@material-ui/icons/AddShoppingCart';
import { makeStyles } from '@material-ui/core/styles';

import { ButtonComponent } from '../_shared/Button';

const useStyles = makeStyles({
  badge: {
    top: '7px',
  },
  image: {
    width: '40px',
  },
  links: {
    textDecoration: 'none',
  },
  icon: {
    color: '#484848',
  },
});

export const Header = ({ isLogged, changeStatus, selectedBooksAmount, removeFavouritesId, removeCartBooks }) => {
  const classes = useStyles();

  const logOut = () => {
    changeStatus(false);
    removeFavouritesId();
    removeCartBooks();
  };

  return (
    <Grid item xs={12}>
      <AppBar position="fixed" color="default">
        <Toolbar disableGutters>
          <Container maxWidth="lg">
            <Grid container>
              <Grid item xs={4} container spacing={2}>
                <Grid item xs={3}>
                  <NavLink exact to="/books" className={classes.links}>
                    <img src="./book.png" alt="logo" className={classes.image} />
                  </NavLink>
                </Grid>
                <Grid item xs={9}>
                  <NavLink exact to="/books" className={classes.links}>
                    <Typography variant="h5" color="primary">
                      GetBooks
                    </Typography>
                  </NavLink>
                </Grid>
              </Grid>
              <Grid item xs={4} />
              <Grid item xs={4} container spacing={2}>
                <Grid item xs={4}>
                  <NavLink exact to="/contacts" className={classes.links}>
                    <ButtonComponent variant="text">Contacts</ButtonComponent>
                  </NavLink>
                </Grid>
                <Grid item xs={4}>
                  <NavLink exact to="/cart" className={classes.links}>
                    <Badge badgeContent={selectedBooksAmount} color="primary" className={classes.badge}>
                      <AddShoppingCartIcon className={classes.icon} />
                    </Badge>
                  </NavLink>
                </Grid>
                <Grid item xs={4}>
                  <NavLink exact to="/" className={classes.links}>
                    <ButtonComponent onClick={logOut} color="primary" disabled={!isLogged}>
                      {isLogged ? 'Log Out' : 'Log In'}
                    </ButtonComponent>
                  </NavLink>
                </Grid>
              </Grid>
            </Grid>
          </Container>
        </Toolbar>
      </AppBar>
      <Toolbar />
    </Grid>
  );
};

Header.propTypes = {
  isLogged: PropTypes.bool.isRequired,
  changeStatus: PropTypes.func.isRequired,
  selectedBooksAmount: PropTypes.number,
  removeFavouritesId: PropTypes.func.isRequired,
  removeCartBooks: PropTypes.func.isRequired,
};

Header.defaultProps = {
  selectedBooksAmount: 0,
};
