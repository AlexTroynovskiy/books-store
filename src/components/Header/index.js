import { connect } from 'react-redux';

import { Header } from './Header';
import { removeFavouritesId } from '../../redux/Books/actions';
import { removeCartBooks } from '../../redux/Cart/actions';

const mapStateToProps = (state) => ({
  selectedBooksAmount: state.cart.books.length,
});

const mapDispatchToProps = {
  removeFavouritesId,
  removeCartBooks,
};

export default connect(mapStateToProps, mapDispatchToProps)(Header);
