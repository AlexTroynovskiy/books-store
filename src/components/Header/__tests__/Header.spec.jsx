import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { Router } from 'react-router';

import HeaderConnect from '../';
import { Header } from '../Header';

describe('Component: Header', () => {
  const initialState = {
    isLogged: false,
    changeStatus: jest.fn(),
    selectedBooksAmount: 0,
    removeFavouritesId: jest.fn(),
  };

  beforeEach(() => {
    jest.resetAllMocks();
  });

  describe('Should be properly connected to the store', () => {
    const mockStore = configureStore();
    let store, wrapper;
    store = mockStore(initialState);

    beforeEach(() => {
      wrapper = shallow(
        <Provider store={store}>
          <HeaderConnect {...initialState} />
        </Provider>
      );
    });

    test('To be in DOM', () => {
      expect(wrapper.exists()).toBe(true);
    });
  });

  describe('To have nested elements', () => {
    let wrapper;
    const historyMock = {
      push: jest.fn(),
      location: {},
      listen: jest.fn(),
      createHref: jest.fn(),
    };

    test('To exist', () => {
      wrapper = shallow(
        <Router history={historyMock}>
          <Header {...initialState} />
        </Router>
      );
      expect(wrapper.exists()).toBe(true);
    });

    test('To match snapshot', () => {
      wrapper = renderer
        .create(
          <Router history={historyMock}>
            <Header {...initialState} />
          </Router>
        )
        .toJSON();
      expect(wrapper).toMatchSnapshot();
    });
  });
});
