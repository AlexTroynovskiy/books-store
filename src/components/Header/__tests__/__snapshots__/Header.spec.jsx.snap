// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Header To have nested elements To match snapshot 1`] = `
<div
  className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
>
  <header
    className="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorDefault mui-fixed MuiPaper-elevation4"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular"
    >
      <div
        className="MuiContainer-root MuiContainer-maxWidthLg"
      >
        <div
          className="MuiGrid-root MuiGrid-container"
        >
          <div
            className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-item MuiGrid-grid-xs-4"
          >
            <div
              className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
            >
              <a
                aria-current={null}
                onClick={[Function]}
              >
                <img
                  alt="logo"
                  src="https://via.placeholder.com/50x40"
                />
              </a>
            </div>
            <div
              className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9"
            >
              <a
                aria-current={null}
                onClick={[Function]}
              >
                <h5
                  className="MuiTypography-root MuiTypography-h5"
                >
                  GetBooks
                </h5>
              </a>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
          />
          <div
            className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-item MuiGrid-grid-xs-4"
          >
            <div
              className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
            >
              <a
                aria-current={null}
                onClick={[Function]}
              >
                <button
                  className="MuiButtonBase-root MuiButton-root MuiButton-text"
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={0}
                  type="button"
                >
                  <span
                    className="MuiButton-label"
                  >
                    Contacts
                  </span>
                </button>
              </a>
            </div>
            <div
              className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
            >
              <a
                aria-current={null}
                onClick={[Function]}
              >
                <span
                  className="MuiBadge-root makeStyles-badge-1"
                >
                  <button
                    aria-label="added to shopping cart"
                    className="MuiButtonBase-root MuiIconButton-root"
                    disabled={false}
                    onBlur={[Function]}
                    onDragLeave={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex={0}
                    type="button"
                  >
                    <span
                      className="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        className="MuiSvgIcon-root"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M11 9h2V6h3V4h-3V1h-2v3H8v2h3v3zm-4 9c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2zm-9.83-3.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.86-7.01L19.42 4h-.01l-1.1 2-2.76 5H8.53l-.13-.27L6.16 6l-.95-2-.94-2H1v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.13 0-.25-.11-.25-.25z"
                        />
                      </svg>
                    </span>
                  </button>
                  <span
                    className="MuiBadge-badge MuiBadge-anchorOriginTopRightRectangle MuiBadge-colorPrimary MuiBadge-invisible"
                  >
                    0
                  </span>
                </span>
              </a>
            </div>
            <div
              className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
            >
              <a
                aria-current={null}
                onClick={[Function]}
              >
                <button
                  className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
                  disabled={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={-1}
                  type="button"
                >
                  <span
                    className="MuiButton-label"
                  >
                    Log In
                  </span>
                </button>
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </header>
  <div
    className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
  />
</div>
`;
